tests/Integration/Behaviour/Features/Context/FeatureFlagFeatureContext.php

```diff
@@ -0,0 +1,159 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace Tests\Integration\Behaviour\Features\Context;
+
+use Doctrine\ORM\EntityManager;
+use Exception;
+use PrestaShopBundle\Entity\FeatureFlag;
+use RuntimeException;
+
+class FeatureFlagFeatureContext extends AbstractPrestaShopFeatureContext
+{
+    /**
+     * "When" steps perform actions, and some of them store the latest result
+     * in this variable so that "Then" action can check its content
+     *
+     * @var mixed
+     */
+    protected $latestResult;
+
+    /**
+     * @Given /^I register a (disabled|enabled) feature flag "(.+)"$/
+     */
+    public function registerFlag(string $state, string $name): void
+    {
+        $doctrineEntityManager = $this->getDoctrineEntityManager();
+
+        try {
+            $newFeatureFlag = new FeatureFlag($name);
+
+            if ($state === 'enabled') {
+                $newFeatureFlag->enable();
+            }
+
+            $doctrineEntityManager->persist($newFeatureFlag);
+            $doctrineEntityManager->flush();
+        } catch (Exception $e) {
+            $this->latestResult = $e;
+        }
+    }
+
+    /**
+     * @When /^I (disable|enable) feature flag "(.+)"$/
+     */
+    public function enableFeatureFlag(string $state, string $name): void
+    {
+        $doctrineEntityManager = $this->getDoctrineEntityManager();
+
+        /** @var FeatureFlag $featureFlag */
+        $featureFlag = $doctrineEntityManager->getRepository('PrestaShopBundle:FeatureFlag')->findOneBy(['name' => $name]);
+
+        if ($state === 'enable') {
+            $featureFlag->enable();
+        } else {
+            $featureFlag->disable();
+        }
+
+        $doctrineEntityManager->flush();
+    }
```

Comment:
IMO status must be a bool and add a toogle.

Reply:
I prefer too and that mains coherent with all toggle.

Reply:
According to "Clean Code: A Handbook of Agile Software Craftsmanship", niladic functions should be preferred to monadic functions.

> The ideal number of arguments for a function is zero (niladic). Next comes one (monadic), followed closely by two (dyadic). Three arguments (triadic) should be avoided where possible. More than three (polyadic) requires very special justification—and then shouldn’t be used anyway.

The less arguments, the less complexity 😄

Reply:
Hi Uncle Bob 😀


The `enableFeatureFlag` method is supposed to enable the flag (but I see it can disable too - if you set a wrong param for exemple :  
```php 
$this->enableFeatureFlag('toto', 'MyFlagName'); // it will disable the flag 😀
```

I suggested the `$state` param should be a boolean and use a status toggler or setter instead of  
```php
   if ($state === 'enable') {
       $featureFlag->enable();
   } else {
       $featureFlag->disable();
   }
```
Something like 
```php
public function changeFeatureFlagState(bool $state, string $name): void
{
      $doctrineEntityManager = $this->getDoctrineEntityManager();

      /** @var FeatureFlag $featureFlag */
      $featureFlag = $doctrineEntityManager
         ->getRepository('PrestaShopBundle:FeatureFlag')
         ->findOneBy(['name' => $name])
         ->setState($state)
       ;

      $doctrineEntityManager->flush();
 }
```

Reply:
Ha I see there is 3 items being discussed here.

> The enableFeatureFlag method

You're right, **this function name is bad because confusing** . I will improve it. ✅ 

> I suggested the $state param should be a boolean

I am not sure I can do that 🤔  . Because `public function enableFeatureFlag(string $state, string $name)` is a function called by Behat, and Behat does mapping between the regexp matched elements and the function arguments.

So as I use the following regexp
```
@Given /^I register a (disabled|enabled) feature flag "(.+)"$/
```

Then Behat matches the `(disabled|enabled)` part with `$state` and the `"(.+)"` with `$name` and since captured elements are strings it gives me strings. I will check but I am not sure Behat call-a-function-using-a-regexp capacities allow using types different than numbers and strings.

> use a status toggler or setter instead of

This is the part where I disagree ^^ and this is why I quoted Uncle Bob. I have multiple reasons why I prefer my implementation.

1. Let's compare `setState($state)` function versus `enable()` and `disable()`

setState($state) | enable() and disable()
----------------|---------------------
when I read it, I cannot know immediately what it does because it depends on $state | when I read it, I immediately know what it does

2. Contrarily to a property like `$name` or `$description` that we expect to be modified in a CRUD fashion without any impact for the system, enabling or disabling a feature flag has impact. Right now you are considering the "CRUD usecase" where we turn feature flags on and off from a BO interface, but this model modification has a real action and meaning. That is why I avoid using `setState()` while I would be happy implementing a `setName()` or `setDescription()`.

3. There are multiple articles that warn us about the dangers of getters and setters: if I quote [this one](https://www.infoworld.com/article/2073723/why-getter-and-setter-methods-are-evil.html):

> Getter and setter methods (also known as accessors) are dangerous for the same reason that public fields are dangerous: They provide external access to implementation details. What if you need to change the accessed field's type? You also have to change the accessor's return type. You use this return value in numerous places, so you must also change all of that code. I want to limit the effects of a change to a single class definition. I don't want them to ripple out into the entire program.

Following this logic, we could one day decide to update the "is enabled / is disabled" implementation, using something else than a status, but this setter will then become irrelevant. We do not have this issue with a `enable()` or `disable()` that simply pretends "I will disable or enable this feature flag, you dont need to know how I will do it".

4. Finally `enable()` and `disable()` have a semantic meaning, which relates to a real-world usecase, while `setState()` has not (for a property like a name, the same logic would make us introduce `rename($name)` function instead of `setName()`).

Reply:
@PululuK @Progi1984 

## About function name

Fixed in https://github.com/PrestaShop/PrestaShop/pull/23650/commits/bab4a0572083f021123539528de102d18074a36c

## About argument type

I confirm this is a limitation of Behat https://docs.behat.org/en/v2.5/guides/2.definitions.html although they dont consider it a limitation

> Regular expression can’t automatically determine the datatype of its matches. So all method arguments coming from step definitions are passed as strings. Even if your regular expression matches “500”, which could be considered an integer, “500” will be passed as a string argument to the step definition’s method.

This is not a feature or limitation of Behat, but rather the inherent way regular expression matching works. It is your responsibility to cast string arguments to integers, floats or booleans where applicable given the code you are testing.

## About `enable/disable` versus `setState`

Nothing to add

Reply:
@matks  merci beaucoup !